G Suite: Add an extra field to collect an `alternateEmail` different from the domain being provisioned

olaseni opened the pull request:
This adds an extra form field to collect a contact email address with a domain different from the domain in the prospective G Suite domain.

#### Changes proposed in this Pull Request

* Added an additional key `alternateEmail` in the default contact fields props.
* Added an additional field to accept an alternate email address with client side validation to keep the associated domain different than the G Suite domain.
* Added an additional key `alternateEmail` in the registration data sent along with the cart for retrieval by the provisioning process.

#### Testing instructions

For the purposes of this test *only*, the domain we want to sample/test e.g `test-domain.com` will be referred to as the domain under test i.e. **DUT**

**Before**

* Load calypso locally ( i.e. calypso.localhost ) and sandbox the API.

* Login to a WP account that has a profile email that ends on the same domain as the DUT. If you have purchased any domains with WP on this account ( i.e registered, transferred ), ensure that the contact email associated with the domain purchase ends with the DUT.

* Attempt to add G Suite to the DUT . You would be presented with a form which collects user registration information.

<img width="741" alt="Screenshot 2019-09-01 at 4 43 14 PM" src="https://user-images.githubusercontent.com/277661/64078863-2a86f280-ccd8-11e9-87f6-85bb6bc04c25.png">

**After**

* Now checkout this branch and allow the Calypso build to be refreshed. Carry out the same action as before the checkout, observe that the registration form now asks for an alternate email address that is validated on the client side. 

* Confirm that the alternate email cannot be invalid or contain a domain that ends in the DUT.

* _You can use a WP account that has an unrelated email in the profile, or domain with unrelated contact email to confirm that the alternate email field is not brought up if the email on file already complies and validates_

<img width="733" alt="Screenshot 2019-09-01 at 4 43 26 PM" src="https://user-images.githubusercontent.com/277661/64078869-35418780-ccd8-11e9-90f3-b2ada07d280b.png">

* This works in harmony with D32310-code

Fixes #

* Prevents an issue where the prospective G Suite account that is about to be created has an email address for the primary contact in the aforementioned domain. This is an error as the email address doesn't exist yet at the time.

olaseni committed:
Fix: added `alternateEmail` to the G Suite user registration data. This ensures that contact email addresses have different domains than the domain about to be provisioned.

This fix introduces:
- An additional key `alternateEmail` in the default contact fields props
- An additional field to accept an alternate email address with client side validation to keep the associated domain different than the G Suite domain
- An additional key `alternateEmail` in the registration data sent along with the cart

a8ci18n commented on client/components/domains/contact-details-form-fields/index.jsx:
```diff
@@ -439,6 +440,17 @@ export class ContactDetailsFormFields extends Component {
+		return (
+			<div className="contact-details-form-fields__row g-suite-alternate-mail">
+				<Input
+					label={ this.props.translate( 'Alternate Email' ) }
```
Hi! I've found a possible matching string that has already been translated **17** times:
`translate( 'Alternate Email Address' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

stephanethomas commented on client/lib/cart-values/cart-items.js:
```diff
@@ -683,6 +683,19 @@ export function fillGoogleAppsRegistrationData( cart, registrationData ) {
+};
+
+export function needsExplicitAlternateEmailForGSuite( cart, contactDetails ) {
+	return some( cart.products, [ 'meta', getDomainPartFromEmail( contactDetails.email ) ] );
```
Shouldn't we display the `Alternate Email` field in the case there is no email address in the contact details?

olaseni replied:
Yes - you're right!

stephanethomas reviewed (changes requested):
Note that changing the email address of the WordPress.com profile is not always enough. If you use a WordPress.com account that was used to purchase a domain, we indeed pick up the email from the domain contact information instead of the email from the WordPress.com profile.

hacchism commented on client/my-sites/checkout/checkout/domain-details-form.jsx:
```diff
@@ -84,14 +87,37 @@ export class DomainDetailsForm extends PureComponent {
+				message = this.props.translate( 'Invalid email address' );
+			} else if ( hasInvalidAlternateEmailDomain( this.props.cart, this.props.contactDetails ) ) {
+				message = this.props.translate(
+					'Email domain cannot be the same as prospective G Suite Domain'
```
+1 to use `Please provide an email address that does not use the same domain than the G Suite account being purchased.`.

> I thought there was a Google support page mentioning this requirement

I tried to locate it and the only support page I could find regarding G Suite billing email address was this one: https://support.google.com/a/answer/1301233?hl=en&ref_topic=2781719. But this page doesn't mention the requirement.

olaseni commented on client/my-sites/checkout/checkout/domain-details-form.jsx:
```diff
@@ -84,14 +87,37 @@ export class DomainDetailsForm extends PureComponent {
+				data = merge( data, { success: false, messages: { alternateEmail: [ message ] } } );
+			}
+		}
+		validationHandler( null, data );
```
Good catch. It is not expected. Thanks.

olaseni commented:
> Note that changing the email address of the WordPress.com profile is not always enough. If you use a WordPress.com account that was used to purchase a domain, we indeed pick up the email from the domain contact information instead of the email from the WordPress.com profile.

Yes I realize. I'll modify the testing instructions to accommodate it.

olaseni committed:
Update: translation corrections and code convention fixes

- Renames methods appropriately.
- Altered translation items based on recommendations.
- Improved the checks on contact email for validation.

stephanethomas reviewed (approved):
This looks good, and works great. I had a minor comment, feel free to merge once it has been addressed.

olaseni committed:
Fix: updated function where email argument could process `undefined`

dmsnell commented on client/lib/cart-values/cart-items.js:
```diff
@@ -683,6 +684,25 @@ export function fillGoogleAppsRegistrationData( cart, registrationData ) {
+		.toLowerCase()
+		.trim()
+		.replace( /.*@([^@]+)$/, '$1' );
+};
```
Hi @olaseni! Nice to meet you. I noticed here that we're applying some unexpected transforms to the email address. I'm not sure what that last `replace` is attempting to do - it seems to acknowledge that multiple `@`s can exist but if it does so, they are in the wrong position, as they would have to fall to the left of the final `@` and also be quoted.

It's good to be pragmatic about the complexity of email addresses, but some times it's easy to be more robust. In this case that might be true. Also, is there a particular reason to force the lower case?

```js
function emailDomain( email ) {
  if ( 'string' !== typeof email || -1 === email.indexOf( '@' ) ) {
    // there's no safe default that makes sense to me
    return null;
  }
  const parts = email.split( '@' );
  const lastPart = parts[ parts.length - 1 ]; // can be multiple @s before domain, but none after split
  return lastPart.replace( />$/, '' ); // could be "Bob <bob@example.com>"
}
```

Anyway, mine is probably also silly. I just figured that when it's easy to be more spec-compliant then it's worth popping in. Saw the PR in my notifications 🤷‍♂

olaseni replied:
Thanks for your observations! As I think about it now I realise that forcing the lower case may be too early, and can be handled later.

The fact that email can also be specified as `Bob <bob@example.com>` would also be considered . 

However the `replace` does exactly what you specify in a different way i.e. it use a capture group in the regex to get the domain after the last `@` character.

I hope I'm picking up what you are putting down.

dmsnell replied:
> However the replace does exactly what you specify in a different way i.e. it uses a regex capture group to grab the string after the last @ character.

Very good! I missed the `$` in the regex when I looked. Might be worth adding a simple explanatory comment…

```js
	.replace( /.*@([^@]+)$/, '$1' ); // domain is everything after last `@`
```

olaseni committed:
Fix: updated code that handles how domain is extracted from code and how domain comparisons are done.

- Improved `getDomainPartFromEmail` to handle more email syntax possibilities.
- Added `isSameDomainAsProductMeta` which handles more concisely the comparison of domains regardless of case.

olaseni committed:
Update: added explanatory comment

olaseni commented on client/lib/cart-values/cart-items.js:
```diff
@@ -683,6 +684,22 @@ export function fillGoogleAppsRegistrationData( cart, registrationData ) {
 }
 
+const getDomainPartFromEmail = emailAddress => {
+	emailAddress = ( `${ emailAddress }` || '' ).toLowerCase().trim();
```
Done
